firedrake/interpolation.py

```diff
@@ -1664,3 +1666,126 @@ def multHermitian(self, source_vec, target_vec):
+
+    See :cite:`Clement:1975` for details.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __new__(cls, expr, V, **kwargs):
+        target_mesh = as_domain(V)
+        source_mesh = extract_unique_domain(expr) or target_mesh
+        if target_mesh is source_mesh:
+            return object.__new__(ClementInterpolator)
+        else:
+            raise ValueError("Clément interpolation requires the source and target meshes to coincide.")
+
+    @no_annotations
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_dofs=False,
+    ):
+        if subset:
+            raise NotImplementedError("subset not implemented")
+        if freeze_expr:
+            raise NotImplementedError("freeze_expr not implemented")
+        if access != op2.WRITE:
+            raise NotImplementedError("access other than op2.WRITE not implemented")
+        if bcs:
+            raise NotImplementedError("bcs not implemented")
+        element = V.ufl_element()
+        if element.family() != "Lagrange" or element.degree() != 1:
+            raise ValueError("Clément interpolation must target a P1 space.")
+        super().__init__(expr, V, subset=subset, freeze_expr=freeze_expr, access=access,
+                         bcs=bcs, allow_missing_dofs=allow_missing_dofs)
+
+    @PETSc.Log.EventDecorator()
+    def interpolate(self, function, output=None, adjoint=False):
+        """
+        Compute the Clément interpolant applied to a source function.
+
+        Parameters
+        ----------
+        function: firedrake.function.Function or firedrake.cofunction.Cofunction
+                  Function to be interpolated.
+        output: firedrake.function.Function or firedrake.cofunction.Cofunction
+                A function to contain the output.
+        adjoint: bool
+                   Set to true to apply the adjoint of the interpolation
+                   operator.
+
+        Returns
+        -------
+        firedrake.function.Function or firedrake.cofunction.Cofunction
+            The resulting interpolated function.
+        """
+        from firedrake.assemble import assemble
+        if adjoint:
+            raise NotImplementedError("Adjoint of Clément interpolation not implemented.")
+        Vs = function.function_space()
+        element = Vs.ufl_element()
+        if not (element.family() == "Discontinuous Lagrange" and element.degree() == 0):
+            raise ValueError("Source function must live in P0 space.")
+        rank = len(Vs.value_shape)
+        if rank not in (0, 1, 2):
+            raise ValueError(f"Rank-{rank} tensors are not supported.")
+        if rank != len(self.V.value_shape):
+            raise ValueError(f"Rank-{rank} input inconsistent with target space.")
+        mesh = self.V.mesh()
+        dim = mesh.topological_dimension()
```

Comment:
This `dim` should be the number of copies of the element `V.block_size`. I think a `VectorFunctionSpace` and `TensorFunctionSpace` could be indexed with a double index `(flat_dof_id, component_id)`. So `rank` could possibly be unrestricted and there wouldn't have to be special cases if you treat the rank-0 as the general case.

Reply:
Addressed in 7ec5197c3f525896e3ce51b9b9bc71dbfb146c00.